_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,202 @@
+
+The benefit of separating listing and printing is quite obvious, because these are in fact completely unrelated tasks - but, to reiterate, it gives us more flexibility in how we want to do any of them (change the formatting without worrying about introducing listing-related bugs), more code reusability (as now we can use the filesystem traversal to, for example, create or delete files along certain paths), ease of testing (as we can test these in isolation) and reasoning (as there is less cognitive overhead when reading code that does _one_ thing instead of _two_ things tangled together). The wonderful nature of functions lets us compose the building blocks back together however we please, to either recover the original intent, or to introduce new functionality.
+
+What we get from separating the tree traversal from listing new branches is, again, quite subtle. First, we know that our traversal works for every structure that at least resembles a tree - and it is ensured by Haskell type system. Second, finding stop condition-related bugs is now easier, as we shifted the responsibility for terminating the recursion to functions that _don't have to be recursive at all!_ Imagine how easy testing becomes!
+
+The beauty of abstract constructs lies in their duality - they are incredibly rigid in one aspect, because their behavior is set in stone and virtually nothing can break it, and at the same time extremely flexible thanks to their generality. Once we create such a function, we can use it with confidence anywhere we wish (provided we can find a use case). The only downside is we need to spend more time to understand what is going on each time we encounter a new function - the more general it is, the harder to grasp it becomes. 
+
+My `treeBuilderM` is, however, far from the most beautiful function. It can get better, prettier and even more general. The forest is thicker and spreads further, with many paths left to explore.
+
+[^excuses]: I am by no means a Haskell expert, as I am still learning the language. Much of the code from this post could be written in a more elegant way, but I either don't know the proper constructs yet, or I am ignoring things like `ifM` on purpose - to make the snippets more accessible to people from outside the Haskell world.
+
+[^axes]:
+    I am consciously ignoring the existence of symbolic links, along with the fact that a given file might not exist - for the sake of brevity.
+    And here is the implementation, based on `listDirectory` and `doesDirectoryExist` from `System.Directory` module, from `directory` package:
+    ```haskell
+    data File = File { base :: Maybe FilePath, name :: FilePath }
+
+    fullPath :: File -> FilePath
+    fullPath (File (Just base) name) = base ++ "/" ++ name
+    fullPath (File Nothing name) = name
+
+    listDir :: File -> IO [File]
```

Comment:
Asking from Haskell-ignorant perspective (so that you can extend your audience to the non-Haskell peoples of the earth): how does `IO X` differ from `IO [X]`? 🤔

Reply:
`[a]` is Haskell syntax for a list of `a`, and `IO [File]` is an IO action returning a list of `File`s.

Reply:
I described `listDir` a bit better to imply what `[]` means.